src/local_cache.cc

```diff
@@ -421,19 +421,18 @@ LocalCache::UpdateLRU(
   // NOTE: Unique lock on cache mutex must be held for this function
 
   const auto& key = cache_iter->first;
-  const auto& cache_entry = cache_iter->second;
-  // Remove key from LRU list if it was already in there
-  auto lru_iter = std::find(lru_.begin(), lru_.end(), key);
-  if (lru_iter != lru_.end()) {
-    lru_.erase(lru_iter);
+  const auto& entry = cache_iter->second;
+  // Remove key from LRU list if entry holds valid LRU iterator already
+  if (entry->lru_iter_.valid_ && entry->lru_iter_.iter_ != lru_.end()) {
```

Comment:
Moving @nnshah1 comment to a thread so it's easier to follow:

> Question: when we evict, we directly pop from the LRU, if we did the same on Insert - could we gurantee the iterator is always valid and avoid the need for a seperate flag? If I understand the flag is only needed once on insert -

My original idea was to just have 2 functions `UpdateLruHit` and `UpdateLruMiss` which would rely on the caller (Lookup/Insert) to know the when the iter is valid or not - and no check/bool would be needed, as you described in your comment. However, I decided to keep it to one function for simplicity, thinking the checks are cheap and would could help catch bugs if changing behavior in the future.

One such future change may be to allow Insert even if a key already exists to overwrite data (perhaps controlled by something in Triton core) when a user/request knows it should invalidate the cached data. In that case, Insert could be on an existing key (that's why I used names `UpdateLruHit/Miss` instead of `UpdateLruLookup/Insert` even though in practice they mean the same thing today).

TLDR; the check isn't absolutely necessary, but I think it helps to add a sanity check. I'm open to breaking this out as described above though. If we want to think about this more, we can always merge this for 23.06 and iterate on it in 23.07.

Reply:
Resolved offline, considering some alternative improvements in the future, but this is good for now.